### ISSUE TITLE
[Fix] too long delay period of Hungary with Scenario.estimate_delay()

### ISSUE THREAD

**Issue by lisphilar** [lisphilar/covid19-sir#693]
## Summary
As discussed in #670, `Scenario.estimate_delay()` returns 53 days as delay period of hungary and this is too long. This causes change points detection failure in S-R trend analysis.

We may need to revise `Scenario.estimate_delay()` to get many data for "Period Length" calculation. We have only 11 records for Hungary as an example.

## Codes
```Python
snl = cs.Scenario(country="Hungary", province=None)
snl.register(jhu_data, population_data, extras=[oxcgrt_data])
# 53, dataframe
delay, df = snl.estimate_delay()
# S-R trend analysis
_ = snl.trend()
```

## Outputs

Index | Confirmed | Stringency_index | Period Length
-- | -- | -- | --
0 | 48757 | 40.74 | NaN
1 | 32 | 50 | 12
2 | 32 | 50 | 36
3 | 32 | 50 | 29
4 | 4183 | 54.63 | 17
5 | 94916 | 57.41 | 53
6 | 3990 | 61.11 | 31
7 | 3741 | 66.67 | 53
8 | 300 | 71.3 | 2
9 | 366279 | 72.22 | 5
10 | 2443 | 76.85 | 118
11 | 614612 | 79.63 | 19

![trend_20_Hungary](https://user-images.githubusercontent.com/7270139/112746449-ce866c00-8fe9-11eb-9444-38c46775ccd3.png)

## Environment

- CovsirPhy version: 2.17.0-pi
- Python version: 3.9.2
- Installation: poetry
- System: WSL (Ubuntu)

**Comment by Inglezos** [comment]
I ran some first tests. I tried the following changes:
In `data_handler.estimate_delay()` I decided to keep `Pelt-rbf` but with `min_size=7`.
In `scenario.estimate_delay()` I used this code:
```
# Filter out long periods
df.dropna(subset=["Period Length"], inplace=True)
df.sort_values("Period Length", inplace=True)
df.reset_index(inplace=True)
Q1 = np.percentile(df["Period Length"], 25, interpolation = 'midpoint') 
Q3 = np.percentile(df["Period Length"], 75, interpolation = 'midpoint')
low_lim = 7 # this is the min_size=7 (default)

# up_lim = Q3 # UNUSED CODE
# df_filtered = df.loc[(df["Period Length"] >= low_lim) & (df["Period Length"] <= up_lim)] # UNUSED CODE

delay_period = int((low_lim + Q1)/2)
```
Something in-between the min default phase duration and Q1. We need something close to 7 but slightly larger.
After trial-error for many countries I concluded to this configuration. We could use another percentile range alternatively or mean/mode/median, even mode.mean/median to the filtered df. I tried this but the values were too large.

If we are confident that at least 7 points can lead to a stable ODE solution, then we don't need to worry about the number of phases.

**Comment by Inglezos** [comment]
Since the `estimate_delay()` is used in the forecast as well, we have to separate the two implementations somehow, probably with a new function? Or we could keep it the same for both. For example, for Greece and Poland, the forecast now using this common `estimate_delay()` seems to be acceptable and does not lead to negative Rt:

![image](https://user-images.githubusercontent.com/36845191/112768691-e0860000-9025-11eb-8277-d561301d5a87.png)
![image](https://user-images.githubusercontent.com/36845191/112768718-06130980-9026-11eb-97b7-d21846f2dbad.png)
![image](https://user-images.githubusercontent.com/36845191/112768701-e976d180-9025-11eb-80e8-8b71886b3288.png)
![image](https://user-images.githubusercontent.com/36845191/112768786-63a75600-9026-11eb-9025-b44ab12085d5.png)

![image](https://user-images.githubusercontent.com/36845191/112768704-ee3b8580-9025-11eb-97b6-9b337fd80ab9.png)
![image](https://user-images.githubusercontent.com/36845191/112768713-fc89a180-9025-11eb-88d1-cfe2937e1eaf.png)
![image](https://user-images.githubusercontent.com/36845191/112768707-f5629380-9025-11eb-9abc-220c7f7f9809.png)
![image](https://user-images.githubusercontent.com/36845191/112768802-76ba2600-9026-11eb-9065-1c1bffbe8588.png)

**Comment by Inglezos** [comment]
We must have in mind the subsequent estimation performance, since the phase separation seems to be critical and what we will decide in this issue will probably affect and #668.
In any case, the suggestions for the forecast issue (bound forecasted parameters to range, keep them positive or even try as target the fraction params_future/params_past) should be implemented and tried independently.

**Comment by lisphilar** [comment]
Thak you for your many tests and ideas!
I will check the phase duration issue with model validator today.

Could you write a pull request to try your ideas? By adding some argumest, including min_size, please do not change the current behaviors and outputs at this time.

**Comment by lisphilar** [comment]
With an easy way using the updated `ModelValidator` class (#695, version 2.18.0-alpha), I tested phase durations = 3, 4, ...9. Smaller number of timepoints showed lower RMSLE scores in this range and this was different from I expected. The method is as follows.

For phase durations = 3, 4, ...9 with SIR/SIR-D/SIR-F model, I repeated the next steps. The eight parameter sets were randomly selcted, but stable for all phase durations because `seed` was not changed.
1. Perform simulation with a specified ODE model
2. Perform parameter estimation
3. Compare the estimated parameters and the parameters produced with th 1st step
4. Repeat trials (1 trial = from the 1st step to the 3rd step)

RMSLE scores are here.
https://gist.github.com/lisphilar/4439bff923358fc6c0451eb7367657b4#file-rmsle_3-9points-csv
(Full: https://gist.github.com/lisphilar/4439bff923358fc6c0451eb7367657b4)

Rows are for trials. Columns are (reproduction number in th 1st column and) the number of timepoints. Values are RMSLE scores. RMSLE with 3 points was lower than that with 4, 5,...9 points.

**Comment by lisphilar** [comment]
Additionally, I tested with phase durations = 10, 30, 50, 70, 90. In this range, the tendency was different from the previous test (3, 4, 5,...9). In the most of trials, minimum RMSLE scores were shown when 50-90 points.

Results are here.
https://gist.github.com/lisphilar/55ad8a76f224d3575dd7429fd3d91856#file-rmsle_10-90points-csv
(Full: https://gist.github.com/lisphilar/55ad8a76f224d3575dd7429fd3d91856)

Comparison of 3 and 10, 90 timepoints is here. There was no clear relationship.
https://gist.github.com/lisphilar/bedc2ec0b5f804f76d2f0a6282649757#file-rmsle_3-10-90points-csv

With these results, RMSLE scores appears to be minimized when phase durations are around 70-90 (but sometimes minimized when 3 timepoints). However, with our observation using S-R trend analysis, it is a difficult solution to set 70-90 as the minimum phase duration.

Steps of one possible solution:
1. perform phase separation with S-R trend analysis and `min_size=7`
2. in each phase found by the 1st step, calculate phase-specific delay period (sorry, details are not clear)
3. when phase duration is shorter than the phase-specific delay period, the phase will be merged to the next phase if this merge has small impact on total score on S-R plane

What do you think about the methods, results and solutions?

**Comment by Inglezos** [comment]
Hmm these results are for sure unexpected, since from what we have seen until now from the various countries analyses, we would expect on the contrary that for increasing `min_size` values (which means larger phase duration) the error would increase.

I agree that we have to keep `min_size` relatively low, ideally between 7-21 days. Would it complicate things if we added some variance with Q1 as I proposed? I will create a test pull request soon for that as you requested.

Sorry but I didn't get the points with "phase-specific delay period" at 2 & 3. What is a phase-specific delay period? And how can we know each one's impact on the whole S-R plane score beforehand?

Another idea would be to use a small `min_size` such as 7 and then continue bottom-up by merging phases depending on the in-between similarity, at the extent that they can be merged while keeping the merged phase's rmsle low. Is this what you mean?

**Comment by Inglezos** [comment]
>This line will be removed if 0.75 for Q2 was added as an argument?
>
I think yes, because Q2 is stricter and more specific than `values < quantile(0.99)`.

I like the idea with phase-specific delay period, and I agree we could either merge phases using this metric or RMSLE, whichever works better.

Sorry for the delayed test pull request, I will try to create it by tomorrow.

**Comment by Inglezos** [comment]
>this period seems unnecessary because we can decide whether merge is required or not, using RMSLE (or the other metrics) scores.
>
Internally in `param_tracker.trend()` if we call `detector.summary(metrics="RMSLE")` for Italy we get:
|      | Start     | End       |   Duration |   RMSLE_S-R |
|:-----|:----------|:----------|-----------:|------------:|
| 0th  | 24Feb2020 | 12Mar2020 |         17 | 1.10819e-05 |
| 1st  | 13Mar2020 | 26Mar2020 |         13 | 2.34475e-05 |
| 2nd  | 27Mar2020 | 11Apr2020 |         15 | 4.88693e-05 |
| 3rd  | 12Apr2020 | 29Apr2020 |         17 | 3.20972e-05 |
| 4th  | 30Apr2020 | 18May2020 |         18 | 1.61877e-05 |
| 5th  | 19May2020 | 03Jun2020 |         15 | 3.71892e-06 |
| 6th  | 04Jun2020 | 22Jun2020 |         18 | 1.93981e-06 |
| 7th  | 23Jun2020 | 07Jul2020 |         14 | 2.45528e-06 |
| 8th  | 08Jul2020 | 21Jul2020 |         13 | 1.60607e-06 |
| 9th  | 22Jul2020 | 02Aug2020 |         11 | 2.06357e-06 |
| 10th | 03Aug2020 | 12Aug2020 |          9 | 1.79075e-06 |
| 11th | 13Aug2020 | 22Aug2020 |          9 | 5.13752e-06 |
| 12th | 23Aug2020 | 01Sep2020 |          9 | 4.15069e-06 |
| 13th | 02Sep2020 | 16Sep2020 |         14 | 7.73165e-06 |
| 14th | 17Sep2020 | 02Oct2020 |         15 | 6.42491e-06 |
| 15th | 03Oct2020 | 12Oct2020 |          9 | 3.0663e-05  |
| 16th | 13Oct2020 | 22Oct2020 |          9 | 3.92305e-05 |
| 17th | 23Oct2020 | 09Nov2020 |         17 | 0.000260708 |
| 18th | 10Nov2020 | 25Nov2020 |         15 | 0.000452957 |
| 19th | 26Nov2020 | 05Dec2020 |          9 | 9.08548e-05 |
| 20th | 06Dec2020 | 15Dec2020 |          9 | 9.54218e-05 |
| 21st | 16Dec2020 | 27Dec2020 |         11 | 3.66871e-05 |
| 22nd | 28Dec2020 | 06Jan2021 |          9 | 6.69033e-05 |
| 23rd | 07Jan2021 | 19Jan2021 |         12 | 9.08299e-05 |
| 24th | 20Jan2021 | 01Feb2021 |         12 | 3.49737e-05 |
| 25th | 02Feb2021 | 11Feb2021 |          9 | 3.5633e-05  |
| 26th | 12Feb2021 | 24Feb2021 |         12 | 7.24826e-05 |
| 27th | 25Feb2021 | 06Mar2021 |          9 | 3.42828e-05 |
| 28th | 07Mar2021 | 16Mar2021 |          9 | 7.39888e-05 |
| 29th | 17Mar2021 | 04Apr2021 |         18 | 0.000118623 |

Here for example what would that threshold be for combining phases (I mean which phases seem suitable to be combined, what is the criteria) and in which function this will be done?

**Comment by lisphilar** [comment]
It is difficult to decide exact value for threshold. One idea is to compare total score (sum of duration * score in each phase) before merging and that after merging. When before > after, merge them. As metric, MSE could be used for total score calculation. RMSLE has "root" and it is not good to sum up.

New method `TrendDetector.combine_auto(metric="MSE")` will be necessary and this will be called by `analysis.ParamTracker.trend()`.

Note:
All available metrics are listed in `util.evaluator.Evaluator.score()`.

**Comment by lisphilar** [comment]
The first problem of this issue was solved and I will close this issue to release the next stable version. We will continue discussion regarding `.estimate_delay()`, `.trend()`, `.estimate()` and `.fit_predict()` in new issues.

**Comment by Inglezos** [comment]
After all these reworks in `estimate_delay()` for the trends analysis, now in scenario.py, the `fit()` method which delay uses? The following one?
![image](https://user-images.githubusercontent.com/36845191/115721413-aee32900-a386-11eb-8ca0-900389766da5.png)

Should we use this indeed or would be more correct in the forecast to use just the `_data-estimate_delay()` as it is? I mean without averaging it with Q1 etc. Or is this happening actually in another function that I don't have in mind right now?

**Comment by lisphilar** [comment]
`Scenario.fit()` uses `int(low_lim + Q1) / 2` because we have `delay, delay_df = self.estimate_delay(oxcgrt_data)` in line 1358 and `handler = RegressionHandler(data=data, model=model, delay=delay, **kwargs)` in line 1374 of scenario.py.

For prediction, can we use hyperparameter optimization or grid search to select the best delay value? The best delay value minimize fitting score calculated by `RegressionHandler`.

**Comment by Inglezos** [comment]
Yes I think that would be a good improvement for forecasting. Perhaps a simple grid search is enough, but maybe with Optuna is more abstract/compact.

**Comment by lisphilar** [comment]
We will move to #731 to change how to calculate delay period in prediction.

**Comment by lisphilar** [comment]
Actually, `.estimate_delay()` may not be the best choice to set `min_size` of S-R trend analysis? Simply, 7 days could be used for `min_size`.

**Comment by Inglezos** [comment]
That's what been bothering me too. It seems that it would be simpler if we just set `min_size` to 7 but I'm not sure if we should not consider at all the `estimate_delay`. It provides some kind of variance and increases the value a little bit which seems to be working well for the phases. After all, the delay is the only measure we have at "application/user level" that represents the actual phases the pandemic is going through, since it indicates the actions, which determine (theoretically) when we have a change of phase or not. Because of this association I thought it would be useful to include delay in the `min_size` value partly.
Do you think that we should simply set it to 7 for simplicity and for the code to be more clear, as the delay was initially created only for the forecast?

**Comment by lisphilar** [comment]
I said `.estimate_delay()` is not necessary for `min_size` calculation because I thought delay period values are notw near to 7. However, my re-checking suggested this is not correct. Also, as you said, the definition of "phase" suggests we need to consider delay period as the minimum size of phase. This logic is not documented, could you document this in TERM.md and some ipynb with a new issue?

My re-checking:
```Python
import covsirphy as cs
data_loader = cs.DataLoader()
jhu_data = data_loader.jhu()
population_data = data_loader.population()
oxcgrt_data = data_loader.oxcgrt()
countries = [
    "Italy", "Japan", "China", "India", "USA",
    "Greece", "Brazil", "Russia", "Spain", "France",
    "United Kingdom", "Germany", "Netherlands", "Czech Republic", "Poland",
    "Sweden", "Austria", "Australia", "Switzerland", "Hungary", "South Africa"
]
delay_values = []
for country in countries:
    snl = cs.Scenario(country=country)
    snl.register(jhu_data, population_data, extras=[oxcgrt_data])
    delay, _ = snl.estimate_delay()
    delay_values.append(delay)

print(delay_values)
```

Output: [8, 7, 7, 9, 8, 9, 8, 7, 7, 9, 9, 7, 11, 10, 10, 7, 9, 7, 9, 9, 8]

**Comment by Inglezos** [comment]
>because I thought delay period values are notw near to 7. However, my re-checking suggested this is not correct. 
>
Yes the delay is relatively close to 7 because we kept Q2 of the normal delay value from `data.estimate_delay()` and then we average this with 7.
I can document this login in TERM.md okay. What ipynb you mean?

**Comment by lisphilar** [comment]
Thank you, "Set phases with S-R trend analysis" in usage_phase.ipynb could be.
https://lisphilar.github.io/covid19-sir/usage_phases.html#Set-phases-with-S-R-trend-analysis
